Act 4
- Natalia Quiroga A01722353
- Paula de Alba A01722262
-  Carolina Treviño García A00835598

In [9]:
import numpy as np
#implementamos la funcion
def f(x, theta):
    a, b, c, d = theta
    return np.cos(a * x) + b * x - c * x**2 + d
#evaluar la solucion
def evaluar_solucion(x, y, theta):
    pred = f(x, theta)
    error = np.abs(pred - y)
    return np.max(error)

def generar_solucion_inicial():
    return np.random.randint(0, 16, size=4)
#generar vecinos
def generar_vecinos(theta):
    vecinos = []
    for i in range(len(theta)):
        if theta[i] > 0:
            vecino = theta.copy()
            vecino[i] -= 1
            vecinos.append(vecino)
        if theta[i] < 15:
            vecino = theta.copy()
            vecino[i] += 1
            vecinos.append(vecino)
    return vecinos
#implementar algortimo de hill climbing
def hill_climbing(x, y, iteraciones=1000):
    theta_actual = generar_solucion_inicial()
    error_actual = evaluar_solucion(x, y, theta_actual)

    for _ in range(iteraciones):
        vecinos = generar_vecinos(theta_actual)
        for vecino in vecinos:
            error_vecino = evaluar_solucion(x, y, vecino)
            if error_vecino < error_actual:
                theta_actual = vecino
                error_actual = error_vecino
    return theta_actual, error_actual

# Datos
x = np.array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0])
y = np.array([1.0, 1.0, 2.0, 4.0, 5.0, 4.0, 4.0, 5.0, 6.0, 5.0])


theta_optimo, error_optimo = hill_climbing(x, y)
print(f"Parámetros óptimos: {theta_optimo}, Error: {error_optimo}")


Parámetros óptimos: [ 8 11  8  1], Error: 1.8508004776987113




*   [7 0 5 5 ], error= 4.715
*   [13 4 7 4], error = 3.597


*   [13 7 12 5], error= 4.847
*   [6 4 10 5], error = 5.125


*   [7 13 12 1], error: 2.289
*   [8 0 0 3] , error: 2.697



*   [8 9 14 5], error: 5.456
*   [8 11 8 1], error: 1.851










El algortimo es una modelacion entre las variables independentienes (x),
y los parametreos (theta) para predecir un valor. La funcion es trigonometrica con polinomios de segundo grado. priemro se genera un vector inicial de parametros (theta) de manera aleatoria en un rango de 0 a 15. Se utiliza el algortimo de hill climbing para iterar los pasos de generacion de vecinos y la evaluacion de soluciones encontradas en el parametro para minimizar el error. el algoritmo se basa en remplazar la solucion actual por una vecina si y solos si la vecina produce un error menor. el proceso se acaba cuando el algortimo encuentra los parametros que resultaron en el menor error encontrado.

los outputs que conseguimos despues de varias iteraciones fueron los que se encuentran desplegados arriba. podemos concluir que le conjunto [8 , 11, 8, 1] fue el conjunto con el menor error de 1.851. dicho esto, es el mejor parametro para modelar la funcion f.

en conclusion, gracias a esete algortimo pudimos ver la optimizacion de ajuste de parametros de un modelo matematico. sin embargo, el hill climbing se puede quedar atrapado en un minimo local y no encontrar el minimo global.